Add owner flag validation during provider enrollment

psekar opened the pull request:
# Summary

***Provide a brief overview of the changes and the issue being addressed. 
This PR addresses the issue #2723 . When enrolling a provider, the owner flag is not validated. Incase the owner flag is invalid or the token has no access to the owner/organization during provider enrollment, the repo registration will fail which is a later step in the process. The change includes checking for a valid GH organization (passed as owner flag) that matches the owner flag and ensure the token has access to the GH organization.

Explain the rationale and any background necessary for understanding the changes. 
List dependencies required by this change, if any.***
I created a type struct to make Git API calls prior to actual provider enrollment flow, also created an interface to be able to mock the tests. The intent of the existing types seems to be useful post provider enrollment so decided to create a new type.

Fixes #(related issue)
#2723 
## Change Type

***Mark the type of change your PR introduces:***

- [x] Bug fix (resolves an issue without affecting existing features)
- [ ] Feature (adds new functionality without breaking changes)
- [ ] Breaking change (may impact existing functionalities or require documentation updates)
- [ ] Documentation (updates or additions to documentation)
- [ ] Refactoring or test improvements (no bug fixes or new functionality)

# Testing

***Outline how the changes were tested, including steps to reproduce and any relevant configurations. 
Attach screenshots if helpful.***
👉 Existing tests will validate the valid owner path.
👉 Added the unit tests to validate the invalid owner flag code path. 
👉  Provide a valid owner flag (a GH org) and provider enrollment is successful
👉  Provide an invalid owner flag and provider enrollment fails during the oauth flow with clear error message about the owner flag.

# Review Checklist:

- [x] Reviewed my own code for quality and clarity.
- [ ] Added comments to complex or tricky code sections.
- [ ] Updated any affected documentation.
- [x] Included tests that validate the fix or feature.
- [ ] Checked that related changes are merged.

eleftherias commented on internal/controlplane/handlers_oauth.go:
```diff
@@ -280,6 +280,22 @@ func (s *Server) processOAuthCallback(ctx context.Context, w http.ResponseWriter
 	if err != nil {
 		return fmt.Errorf("error encoding token: %w", err)
 	}
+ 	login, err := s.ghLite.GetLoginByToken(ctx, token);
+	if (err != nil) {
+		return fmt.Errorf("error getting login: %w", err)
+	}
+	// Verify if the token user is a member of the ownerFilter (Organization)
+	if (stateData.OwnerFilter.Valid) {
+		member, err := s.ghLite.IsMemberInOrganization(ctx, token, stateData.OwnerFilter.String, login)
```
What do you think about using the [List org memberships for authenticated users](https://docs.github.com/en/rest/orgs/members?apiVersion=2022-11-28#list-organization-memberships-for-the-authenticated-user) API instead? Then we would only need 1 API call instead of 2.

psekar replied:
Good catch. Yes this API serves the purpose here.

eleftherias reviewed (changes requested):
Thanks for the contribution @psekar!

This approach looks good to me! I've added a few comments inline with suggestions that will simplify the code. 
Feel free to reach out if you have any questions.

dmjb commented on README.md:
````diff
@@ -206,6 +206,12 @@ See the documentation on how to do that - [Docs](https://minder-docs.stacklok.de
+For local development, create the token_key_passphrase in .ssh folder. For production, these tokens can be served from a secure vault.
+
+```bash
+openssl rand -base64 32 > .ssh/token_key_passphrase
````
This is typically carried out by using the `make bootstrap` command. I would suggest that you document the make command instead of the individual steps it carries out.

psekar replied:
Sure 👍

psekar commented on internal/providers/github/common.go:
```diff
@@ -86,6 +86,11 @@ type GitHub struct {
 // Ensure that the GitHub client implements the GitHub interface
 var _ provifv1.GitHub = (*GitHub)(nil)
 
+type GithubLite struct {
```
yea, this makes sense. Thanks for pointing it out.

psekar commented on internal/controlplane/handlers_githubwebhooks_test.go:
```diff
@@ -117,7 +117,7 @@ func (s *UnitTestSuite) TestHandleWebHookPing() {
 
 	mockStore := mockdb.NewMockStore(ctrl)
-	srv, evt := newDefaultServer(t, mockStore, nil)
+	srv, evt := newDefaultServer(t, mockStore, nil, true)
```
I am mocking the behavior in the test but i am ghProviderService as an argument so other tests can still get a non mocked version of GhProviderService. Let me know if this makes sense.

eleftherias commented on internal/controlplane/handlers_oauth_test.go:
```diff
@@ -518,6 +527,18 @@ func TestProviderCallback(t *testing.T) {
+		},
+		code: 403,
+		err: "The ownerFilter value provided does not allow access for the logged in user\n",
+		buildStubs: func(service *mockprovsvc.MockGitHubProviderService) {
```
For consistency could this be a function `withInvalidOrgMemberships`

eleftherias commented on internal/controlplane/handlers_oauth_test.go:
```diff
@@ -497,6 +504,7 @@ func TestProviderCallback(t *testing.T) {
 		},
 		code: 403,
 		err:  "The provided login token was associated with a different user.\n",
+		buildStubs: nil,
```
This should be a no-op function. Setting it to nil causes a nil pointer dereference error.

eleftherias commented on internal/controlplane/handlers_oauth_test.go:
```diff
@@ -572,7 +593,9 @@ func TestProviderCallback(t *testing.T) {
-			s, _ := newDefaultServer(t, store, clientFactory)
+			mockGhProviderService := mockprovsvc.NewMockGitHubProviderService(ctrl)
+			tc.buildStubs(mockGhProviderService)
+			s, _ := newDefaultServer(t, store, clientFactory, mockGhProviderService)
```
Instead of changing the signature of `newDefaultServer` let's set `s.ghProviders = mockGhProviderService` after creating the server. That's what we do for `providerAuthManager`, `sessionService` etc.

psekar replied:
I do see this pattern in other places too after your comment. I tried not to change the server behavior outside of server_test but then we create a new instance so it shouldnt matter. Thanks for pointing it out 👍

eleftherias reviewed (changes requested):
Thanks for the updates @psekar! 

I've added a few follow-up comments.
You can also run `make lint` before committing to makes sure the code passes the lint rules.

psekar commented:
> Thanks for the updates @psekar!
> 
> I've added a few follow-up comments. You can also run `make lint` before committing to makes sure the code passes the lint rules.

I am still very new to Go and your review has been super useful 🙏

eleftherias commented on internal/controlplane/handlers_oauth.go:
```diff
@@ -290,6 +290,18 @@ func (s *Server) processOAuthCallback(ctx context.Context, w http.ResponseWriter
 	}
+	// Verify if the token user is a member of the ownerFilter (Organization)
+	if stateData.OwnerFilter.Valid {
+		member, err := s.ghProviders.ValidateOrgMembershipForToken(ctx, token, stateData.OwnerFilter.String)
+		if err != nil {
+			return fmt.Errorf("error checking organization membership: %w", err)
+		}
+
+		if !member {
+			return newHttpError(http.StatusForbidden, "Invalid OwnerFilter Provided").SetContents(
+				"The ownerFilter value provided does not allow access for the logged in user")
+		}
```
You may have noticed that the linter is complaining that the [cyclomatic complexity](https://en.wikipedia.org/wiki/Cyclomatic_complexity) of this function is too high.

Essentially, this means this function has several code paths which makes it hard to reason about when reading through it.

We can fix this by extracting the block here into a separate `validateOwnerFilter` function:
```golang
if stateData.OwnerFilter.Valid {
	err := s.validateOwnerFilter(ctx, token, stateData)
	if err != nil {
		return err
	}
}
```

psekar replied:
I fixed the 2 lint issues that i overlooked in my last review. Thanks

eleftherias reviewed (changes requested):
Thanks for the quick updates again @psekar!

I've left some final comments about cleanup / linting changes.

eleftherias reviewed (commented):
I also noticed that the mocks need to be updated. You can run `make gen` in the root directory before committing your code and that will update them.

psekar commented on internal/controlplane/handlers_oauth.go:
```diff
@@ -290,6 +290,18 @@ func (s *Server) processOAuthCallback(ctx context.Context, w http.ResponseWriter
+
+		if !member {
+			return newHttpError(http.StatusForbidden, "Invalid OwnerFilter Provided").SetContents(
+				"The ownerFilter value provided does not allow access for the logged in user")
```
After the refactor, I noticed we will be returning the same error message incase of invalid org or any github API failures. Do you think we should handle the github API failure separately ?

eleftherias replied:
Instead of returning the boolean from the helper function, we can just return the error. Then the caller can directly return the error as well. This way, we can keep the error messages for invalid org or GitHub API failures.
It's fairly common to do that in our codebase, for example:
https://github.com/stacklok/minder/blob/16421648e284c9f5be7c5261aaa8e5a94b61c70c/internal/controlplane/handlers_oauth.go#L345-L348

psekar commented:
> [![Coverage Status](https://camo.githubusercontent.com/13636dc75840580a091f7ba522d772b6898c2359183bee6553db53b021c24d97/68747470733a2f2f636f766572616c6c732e696f2f6275696c64732f36393138393739322f6261646765)](https://coveralls.io/builds/69189792)
> 
> coverage: 53.872% (-0.03%) from 53.899% when pulling **[94472cb](https://github.com/stacklok/minder/commit/94472cb57df55af1bb0091a5f59563bb9a55c74b) on tinytrail:validate-owner-filter** into **[e438848](https://github.com/stacklok/minder/commit/e438848a25bacfedc538d5f604e3490ec48b1683) on stacklok:main**.

Will add more test coverage and send it back for review tomorrow.

psekar commented:
> > [![Coverage Status](https://camo.githubusercontent.com/13636dc75840580a091f7ba522d772b6898c2359183bee6553db53b021c24d97/68747470733a2f2f636f766572616c6c732e696f2f6275696c64732f36393138393739322f6261646765)](https://coveralls.io/builds/69189792)
> > coverage: 53.872% (-0.03%) from 53.899% when pulling **[94472cb](https://github.com/stacklok/minder/commit/94472cb57df55af1bb0091a5f59563bb9a55c74b) on tinytrail:validate-owner-filter** into **[e438848](https://github.com/stacklok/minder/commit/e438848a25bacfedc538d5f604e3490ec48b1683) on stacklok:main**.
> 
> Will add more test coverage and send it back for review tomorrow.

Fixed the linter issues and added more test coverage now. @eleftherias Appreciate your review 👍

eleftherias commented on internal/controlplane/handlers_oauth.go:
```diff
@@ -291,6 +291,19 @@ func (s *Server) processOAuthCallback(ctx context.Context, w http.ResponseWriter
+		if (err != nil) {
+			return err
+		}
+		if !member {
+			return newHttpError(http.StatusForbidden, "Invalid OwnerFilter Provided").SetContents(
+				"You do not have access to the organization %s, specified in the owner filter."+
+					" Please ensure that you are a member of the GitHub organization and try again.", stateData.OwnerFilter.String)
+		}
+	}
```
We don't need to return the `member` boolean from the helper, we can simply return the error, and then this second `if` statement is not needed.

The helper would be:
```
func (s *Server) validateOwnerFilter(ctx context.Context, token *oauth2.Token, ownerFilter string) error {
	member, err := s.ghProviders.ValidateOrgMembershipForToken(ctx, token, ownerFilter)
	if err != nil {
		return fmt.Errorf("error checking organization membership: %w", err)
	}

	if !member {
		return newHttpError(http.StatusForbidden, "Invalid OwnerFilter Provided").SetContents(
			"You do not have access to the organization %s, specified in the owner filter."+
				" Please ensure that you are a member of the GitHub organization and try again.", ownerFilter)
	}
	return nil
}
```

psekar replied:
I recall running `make lint` before i sent the change to upstream, but i might have missed running it after the last refactor. Thanks for catching this.

eleftherias reviewed (changes requested):
Thank you for your patience on this @psekar!
The cyclomatic complexity of `processOAuthCallback` is still high, I've proposed a change inline that will fix it. After this change, we should be ready to merge.
